File: public_web/tests/vixrex-dogal-netlestirme-kabul.test.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";
import {
  VIXREX_DOGAL_GENEL_SORU,
  vixrexBaglamsalCevapKarari,
  type VixrexBaglamKarari,
  type VixrexBekleyenBaglam,
} from "../src/lib/vixrexConversationContext";

interface Senaryo {
  id: string;
  bekleyen: VixrexBekleyenBaglam | null;
  girdi: string;
  karar: VixrexBaglamKarari;
  yazma: boolean;
}

const senaryolar = JSON.parse(
  readFileSync(
    resolve(__dirname, "../../shared/vixrex_dogal_netlestirme_senaryolari.json"),
    "utf8",
  ),
) as Senaryo[];

describe("Vixrex doğal netleştirme bağımsız kabul kümesi", () => {
  it("en az 20 konuşma senaryosu içerir", () => {
    expect(senaryolar.length).toBeGreaterThanOrEqual(20);
  });

  for (const senaryo of senaryolar) {
    it(`${senaryo.id}: bağlama göre yazma/sorma kararını verir`, () => {
      const sonuc = vixrexBaglamsalCevapKarari(senaryo.girdi, senaryo.bekleyen);
      expect(sonuc.karar).toBe(senaryo.karar);
      expect(sonuc.yazma).toBe(senaryo.yazma);
      if (!senaryo.yazma) expect(sonuc.mesaj.trim().length).toBeGreaterThan(0);
    });
  }

  it("bağlam yokken kullanıcıdan sistemin iç alan adını istemez", () => {
    const sonuc = vixrexBaglamsalCevapKarari("bunu düzelt", null);
    expect(sonuc.mesaj).toBe(VIXREX_DOGAL_GENEL_SORU);
    expect(sonuc.mesaj.toLocaleLowerCase("tr-TR")).not.toContain("hangi alan");
  });
});

File: test/vixrex_dogal_netlestirme_kabul_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:vixrex/services/vixrex_nlu/vixrex_conversation_context.dart';

String kararAdi(VixrexBaglamKarari karar) => switch (karar) {
  VixrexBaglamKarari.genelSor => 'genel_sor',
  VixrexBaglamKarari.ozelSor => 'ozel_sor',
  VixrexBaglamKarari.ayniKalsin => 'ayni_kalsin',
  VixrexBaglamKarari.kaldirmaOnayi => 'kaldirma_onayi',
  VixrexBaglamKarari.kaldir => 'kaldir',
  VixrexBaglamKarari.iptal => 'iptal',
  VixrexBaglamKarari.boolTrue => 'bool_true',
  VixrexBaglamKarari.boolFalse => 'bool_false',
  VixrexBaglamKarari.saatEksik => 'saat_eksik',
  VixrexBaglamKarari.adresEksik => 'adres_eksik',
  VixrexBaglamKarari.deger => 'deger',
};

void main() {
  final raw =
      jsonDecode(
            File(
              'shared/vixrex_dogal_netlestirme_senaryolari.json',
            ).readAsStringSync(),
          )
          as List<dynamic>;

  group('Vixrex doğal netleştirme bağımsız kabul kümesi', () {
    test('en az 20 konuşma senaryosu içerir', () {
      expect(raw.length, greaterThanOrEqualTo(20));
    });

    for (final item in raw.cast<Map<String, dynamic>>()) {
      test('${item['id']}: Flutter aynı bağlam kararını verir', () {
        final bekleyenJson = item['bekleyen'] as Map<String, dynamic>?;
        final bekleyen =
            bekleyenJson == null
                ? null
                : VixrexBekleyenBaglam(
                  anahtar: bekleyenJson['anahtar'] as String,
                  etiket: bekleyenJson['etiket'] as String,
                  tip: bekleyenJson['tip'] as String,
                  eylem: bekleyenJson['eylem'] as String?,
                );
        final sonuc = vixrexBaglamsalCevapKarari(
          item['girdi'] as String,
          bekleyen,
        );

        expect(kararAdi(sonuc.karar), item['karar']);
        expect(sonuc.yazma, item['yazma']);
        if (!(item['yazma'] as bool)) {
          expect(sonuc.mesaj.trim(), isNotEmpty);
        }
      });
    }

    test('bağlam yokken kullanıcıdan sistemin iç alan adını istemez', () {
      final sonuc = vixrexBaglamsalCevapKarari('bunu düzelt', null);
      expect(sonuc.mesaj, vixrexDogalGenelSoru);
      expect(sonuc.mesaj.toLowerCase(), isNot(contains('hangi alan')));
    });
  });
}
